### DISCUSSION TITLE
Möglichkeit herauszufinden wann der Wechselrichter das letztemal erfolgreich ausgelesen wurde?

### DISCUSSION THREAD

**Discussion by SofieF2005** [weltenwort/home-assistant-rct-power-integration#359] [Ideas]
Hallo. Ich stehe aktuell vor dem Problem dass man in HA nicht so einfach herausfinden kann wann die Daten das letzte mal geupdatet wurden wenn die Werte sich nicht grad ändern. Wenn HA die Verbindung zum Wechselrichter verliert äußert sich das auch nur so dass nichts mehr geupdatet wird. Ich habe aber keine Entity gefunden wodurch ich dieses einfach checken könnte.
An meinen MQTT Broker sende ich alle paar Sekunden auch die Werte weiter. Auch hier werden munter dann zyklisch eben fortan die immer gleichen Werte geschickt von beispielsweise 503W Hausverbrauch auch wenn der Wechselrichter gar keine Verbindung mehr hat.
Gibt es irgendwie eine Möglichkeit "irgendetwas" zu implementieren vielleicht einen Counter der hochläuft oder einen Zeitstempel der sich bei jedem auslesen aktualisiert woran man dann in HA erkennen würde wie aktuell die Daten tatächlich sind?
Ein Zähler der bei jedem Auslesen seinen Wert verändert würde schon vieles einfacher machen. Zwar ändern sich gewisse Verbräuche ständig aber eine Erkennung darüber ist dennoch etwas unsauber weil es rein theoretisch auch mal sein kann dass über einen Zeitraum sich eben nichts ändert.
Eine Entity "updated" mit nem Zeitstempel wäre sehr praktisch. Und/Oder eine Entity welche 0 bzw 1 zurückgibt davon abhängig ob der Inverter noch erreichbar ist oder nicht

**Comment by weltenwort** [comment]
Hi und danke für den Vorschlag :wave: Auch wenn für mich die Verbindung recht zuverlässig funktioniert kann ich den Bedarf dennoch nachvollziehen. Das Problem mit einem Counter oder einem timestamp Attribut wäre dass es deutlich mehr Daten erzeugen würde, was die Datenbank aufbläht.

Was aber vielleicht das Problem auch lösen könnte wäre ein verbesserte Feststellung des "unavailable" Status für die Entities. So könnte man z.B. die Entity auf "unavailable" setzen wenn das abrufen 2 oder 3 mal hintereinander fehlgeschlagen ist. Dies könnte man dann mit einer Automatisierung erkennen und eine Notification oder so absetzen. Klingt das sinnvoll?

**Reply by SofieF2005** [reply]
Hallo, nein ich dachte nicht an ein Attribut für einen Timestamp für jede Entität sondern als eine einzige Entität welche als Inhalt den Timestamp hat wann der Wechselrichter zuletzt erfolgreich ausgelesen wurde. So könnte man in HomeAssistant einfach auf Wertänderung prüfen weil dieser Wert sich ja auf jedenfall ändern wird auch wenn beispielsweise Leistungen und Temperaturen konstant bleiben weil sich manche ggf nur langsam ändert wie SOC.

Ich vermute mal dein Script baut in Regelmäßigen Abständen eine Verbindung auf und fragt dann die Werte alle ab die es haben will. Die erfolgreiche übermittlung der Daten könnte man dann als Timestamp benutzen.
Im Grunde würde da auch nen Zähler reichen. Geht ja einfach nur darum unterscheiden zu können 

Ja mit unavaible sollte das ggf auch klappen denke ich. Wobei ich unavaible sporadisch auch von Zeit zu Zeit bekomme auf immer unterschiedlichen Entitäten, ich denke aber das ist normal?

**Reply by weltenwort** [reply]
Ja, eigentlich ist "unavailable" die offizielle Art wie Home Assistant eine Nichterreichbarkeit des Geräts darstellt. Ich denke es müsste auch mindestens drei mal hintereinander die Kommunikation fehlschlagen damit eine Entität den Status annimmt. Und wenn man dann eine Automatisierung darauf aufbaut würde man ja vmtl. sowieso den state trigger mit einer Dauer kombinieren.

**Reply by SofieF2005** [reply]
Aber unavaible sagt er nicht wenn man den Stecker zieht. Ich glaube das hängt eher mit dem Wechselrichter zusammen. Das ist bei der evcc Integration bei diesem Wechselrichter auch so. Er kann nur eine Verbindung aufbauen und wenn man dann gleichzeitig die Hersteller App öffnet häufigen sich die Fehlermeldungen weil die App auch gleichzeitig versucht zu lesen. (Aus dem Gedächtnis heraus).
Da ich die Daten universell nutzen will habe ich ihn daher in HA integriert und sende die Werte per MQTT weiter an Evcc. Evcc ist so eingestellt per default dass es spätestens alle 30sekunden neue Werte per MQTT erwartet und ansonsten anzeigt dass der Wert zu alt ist. Das ist ja auch mein Problem, wenn ich zyklisch sende trickse ich diese Fehlererkennung in evcc quasi aus weil im Falle eines Verbindungsverlust mqtt dann einfach weiterhin den letzten Wert sendet selbst wenn dieser schon 2h alt wäre...
"Send by change" funktioniert zwar bei Entitäten die sich alle paar Sekunden ändern, aber weniger bei so etwas wie dem SOC. Davon ab wäre das auch nur ein Workarround in meinen Augen und nicht "sauber".
Daher sollten auch hier die Werte irgendwie als "out of date" erkannt werden können wenn zu alt weil zB keine Verbindung zum WR mehr besteht.

Man läd sonst beispielsweise sein Fahrzeug weil die Logik denkt es ist noch genug PV überschuss da, dabei ist die aktuelle Leistung schon ganz woanders nur weil immer die gleichen alten Werte zyklisch gesendet werden.

Problem ist auch HA selber irgendwo

https://community.home-assistant.io/t/add-a-last-refreshed-to-complement-last-changed-and-last-updated/620462

**Reply by weltenwort** [reply]
Ja, verstehe. Aber das klingt doch so als müsste "unavailable" für eine Bedingung ausreichen, damit die Automation erkennen kann dass kein aktueller Wert vorhanden ist, oder?

**Reply by SofieF2005** [reply]
ja aber aktuell geht er wenn man die Verbindung trennt nicht auf unavaible sondern behält den letzten Wert bei.

**Reply by weltenwort** [reply]
Genau, das würde ich ändern :+1:

**Comment by weltenwort** [comment]
Assuming it would work for your use-case, I filed https://github.com/weltenwort/home-assistant-rct-power-integration/issues/360 track that request.
